### ISSUE TITLE
Function deployment fails in a single-node vhive cluster

### ISSUE THREAD

**Issue by wangtianxia-sjtu** [vhive-serverless/vHive#702]
I have been following the [guide](https://github.com/vhive-serverless/vHive/blob/main/docs/quickstart_guide.md) to start a single node vhive cluster. However, when I am deploying the function, there are some unexpected failures.

```
$ source /etc/profile && pushd ./examples/deployer && go build && popd && ./examples/deployer/deployer
~/vHive/examples/deployer ~/vHive ~/vHive
~/vHive ~/vHive
WARN[0004] Failed to deploy function pyaes-1, configs/knative_workloads/pyaes.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0004] Deployed function pyaes-1
WARN[0004] Failed to deploy function rnn-serving-1, configs/knative_workloads/rnn_serving.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0004] Deployed function rnn-serving-1
WARN[0004] Failed to deploy function helloworld-0, configs/knative_workloads/helloworld.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0004] Deployed function helloworld-0
WARN[0004] Failed to deploy function pyaes-0, configs/knative_workloads/pyaes.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0004] Deployed function pyaes-0
WARN[0004] Failed to deploy function rnn-serving-0, configs/knative_workloads/rnn_serving.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0004] Deployed function rnn-serving-0
WARN[0008] Failed to deploy function rnn-serving-2, configs/knative_workloads/rnn_serving.yaml: exit status 1
Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
Run 'kn --help' for usage

INFO[0008] Deployed function rnn-serving-2
INFO[0008] Deployment finished
```

It seems that the `kn` client cannot resolve `webhook.knative-serving.svc`. I cannot ping `webhook.knative-serving.svc`, either.

```
$ ping webhook.knative-serving.svc
ping: webhook.knative-serving.svc: Name or service not known
```

Maybe there is something I have missed, but can anyone help?

**Comment by wangtianxia-sjtu** [comment]
> I have been following the [guide](https://github.com/vhive-serverless/vHive/blob/main/docs/quickstart_guide.md) to start a single node vhive cluster. However, when I am deploying the function, there are some unexpected failures.
> 
> ```
> $ source /etc/profile && pushd ./examples/deployer && go build && popd && ./examples/deployer/deployer
> ~/vHive/examples/deployer ~/vHive ~/vHive
> ~/vHive ~/vHive
> WARN[0004] Failed to deploy function pyaes-1, configs/knative_workloads/pyaes.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0004] Deployed function pyaes-1
> WARN[0004] Failed to deploy function rnn-serving-1, configs/knative_workloads/rnn_serving.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0004] Deployed function rnn-serving-1
> WARN[0004] Failed to deploy function helloworld-0, configs/knative_workloads/helloworld.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0004] Deployed function helloworld-0
> WARN[0004] Failed to deploy function pyaes-0, configs/knative_workloads/pyaes.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0004] Deployed function pyaes-0
> WARN[0004] Failed to deploy function rnn-serving-0, configs/knative_workloads/rnn_serving.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0004] Deployed function rnn-serving-0
> WARN[0008] Failed to deploy function rnn-serving-2, configs/knative_workloads/rnn_serving.yaml: exit status 1
> Error: Internal error occurred: failed calling webhook "webhook.serving.knative.dev": failed to call webhook: Post "https://webhook.knative-serving.svc:443/defaulting?timeout=10s": EOF
> Run 'kn --help' for usage
> 
> INFO[0008] Deployed function rnn-serving-2
> INFO[0008] Deployment finished
> ```
> 
> It seems that the `kn` client cannot resolve `webhook.knative-serving.svc`. I cannot ping `webhook.knative-serving.svc`, either.
> 
> ```
> $ ping webhook.knative-serving.svc
> ping: webhook.knative-serving.svc: Name or service not known
> ```
> 
> Maybe there is something I have missed, but can anyone help?

The problem above is already solved.

However, when I am invoking the function with `pushd ./examples/invoker && go build && popd && ./examples/invoker/invoker`, all the invocations are not completed.

```
~/vHive/examples/invoker ~/vHive
~/vHive
INFO[2023-03-21T05:57:25.772255511Z] Reading the endpoints from the file: endpoints.json
INFO[2023-03-21T05:57:30.772633619Z] Issued / completed requests: 5, 0
INFO[2023-03-21T05:57:30.772765745Z] Real / target RPS: 0.00 / 1
INFO[2023-03-21T05:57:30.772785288Z] Experiment finished!
INFO[2023-03-21T05:57:30.772802660Z] The measured latencies are saved in rps0.00_lat.csv
```

If I directly access the function URL with wget, it will report `502 Bad Gateway`.

```
$ wget http://helloworld-0.default.192.168.1.240.sslip.io
--2023-03-21 06:34:20--  http://helloworld-0.default.192.168.1.240.sslip.io/
Resolving helloworld-0.default.192.168.1.240.sslip.io (helloworld-0.default.192.168.1.240.sslip.io)... 192.168.1.240
Connecting to helloworld-0.default.192.168.1.240.sslip.io (helloworld-0.default.192.168.1.240.sslip.io)|192.168.1.240|:80... connected.
HTTP request sent, awaiting response... 502 Bad Gateway
2023-03-21 06:34:20 ERROR 502: Bad Gateway.
```

**Comment by ustiugov** [comment]
our functions are gRPC servers not HTTP servers, you need to use grpcurl or the invoker to trigger them. some of our CI workflow use grpcurl, you can take a look 

I hope this helps, you can resolve the issue if it's fixed.

**Comment by CuriousGeorgiy** [comment]
Please attach logs of the whole setup process, not only function deployment.

> The problem above is already solved.

How did you solve it or make a conclusion that it got solved?

**Comment by wangtianxia-sjtu** [comment]
Sorry for the late response. This is how I setup a single node cluster. The problem is that no request sent by the `./examples/invoker/invoker` is completed. I have attached all the logs below.

PS: The previous problem is caused by the proxy settings. The access to `webhook.knative-serving.svc` should not be intercepted by the proxy. Since I must have set a http_proxy variable to access images on `gcr.io`, I need to specify with a `no_proxy` that some domains (e.g.: `webhook.knative-serving.svc`) should not be accessed with the proxy. I solve the problem by setting `no_proxy` environment variable.

# Setup vhive

```bash
mkdir -p /tmp/vhive-logs
```

Run the setup scripts of the master node

```bash
./scripts/cloudlab/setup_node.sh > >(tee -a /tmp/vhive-logs/setup_node.stdout) 2> >(tee -a /tmp/vhive-logs/setup_node.stderr >&2)
```

[setup_node.stderr.log](https://github.com/vhive-serverless/vHive/files/11068544/setup_node.stderr.log)
[setup_node.stdout.log](https://github.com/vhive-serverless/vHive/files/11068545/setup_node.stdout.log)

Run the setup scripts of the worker node

```bash
./scripts/cluster/setup_worker_kubelet.sh > >(tee -a /tmp/vhive-logs/setup_worker_kubelet.stdout) 2> >(tee -a /tmp/vhive-logs/setup_worker_kubelet.stderr >&2)
```

[setup_worker_kubelet.stderr.log](https://github.com/vhive-serverless/vHive/files/11068546/setup_worker_kubelet.stderr.log)
[setup_worker_kubelet.stdout.log](https://github.com/vhive-serverless/vHive/files/11068547/setup_worker_kubelet.stdout.log)

Run the `containerd` command in one tmux window.

`sudo -E` is to pass the `http_proxy` variable to the `containerd` process.

```bash
sudo -E containerd > >(tee -a /tmp/vhive-logs/containerd.stdout) 2> >(tee -a /tmp/vhive-logs/containerd.stderr >&2)
```

[containerd.stderr.log](https://github.com/vhive-serverless/vHive/files/11068538/containerd.stderr.log)
[containerd.stdout.log](https://github.com/vhive-serverless/vHive/files/11068539/containerd.stdout.log)

Run the `firecracker-containerd` in another tmux window.

```bash
sudo -E /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml > >(tee -a /tmp/vhive-logs/firecracker.stdout) 2> >(tee -a /tmp/vhive-logs/firecracker.stderr >&2)
```

[firecracker.stderr.log](https://github.com/vhive-serverless/vHive/files/11068542/firecracker.stderr.log)
[firecracker.stdout.log](https://github.com/vhive-serverless/vHive/files/11068543/firecracker.stdout.log)

Run the `vhive` in the third tmux window.

```bash
sudo -E ./vhive > >(tee -a /tmp/vhive-logs/vhive.stdout) 2> >(tee -a /tmp/vhive-logs/vhive.stderr >&2)
```

[vhive.stderr.log](https://github.com/vhive-serverless/vHive/files/11068548/vhive.stderr.log)
[vhive.stdout.log](https://github.com/vhive-serverless/vHive/files/11068550/vhive.stdout.log)

Run the scripts to create a single node cluster

```bash
./scripts/cluster/create_one_node_cluster.sh > >(tee -a /tmp/vhive-logs/create_cluster.stdout) 2> >(tee -a /tmp/vhive-logs/create_cluster.stderr >&2)
```

[create_cluster.stderr.log](https://github.com/vhive-serverless/vHive/files/11068540/create_cluster.stderr.log)
[create_cluster.stdout.log](https://github.com/vhive-serverless/vHive/files/11068541/create_cluster.stdout.log)

However, there are some errors here in the `create_cluster.stderr.log`
```
[WARNING CRI]: container runtime is not running: output: time="2023-03-25T09:29:39Z" level=fatal msg="validate service connection: CRI v1 runtime API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.23.17: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/kube-controller-manager:v1.23.17: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/kube-scheduler:v1.23.17: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/kube-proxy:v1.23.17: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/pause:3.6: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/etcd:3.5.1-0: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
	[WARNING ImagePull]: failed to pull image k8s.gcr.io/coredns/coredns:v1.8.6: output: time="2023-03-25T09:29:40Z" level=fatal msg="validate service connection: CRI v1 image API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.ImageService"
, error: exit status 1
```

# Deploy functions

examples/deployer/functions.json:

```
{
    "functions": [
        {
            "name": "helloworld",
            "file": "helloworld.yaml",
            "count": 1
        }
    ]
}
``` 

deploy client:

```bash
$ source /etc/profile && pushd ./examples/deployer && go build && popd && ./examples/deployer/deployer
~/vHive/examples/deployer ~/vHive
~/vHive
INFO[0025] Deployed function helloworld-0
INFO[0025] Deployment finished
```

Check the deployment:

```
$ kn service list
NAME           URL                                                  LATEST               AGE     CONDITIONS   READY   REASON
helloworld-0   http://helloworld-0.default.192.168.1.240.sslip.io   helloworld-0-00001   5m56s   3 OK / 3     True
```

Invoke the function:

```
$ pushd ./examples/invoker && go build && popd && ./examples/invoker/invoker
~/vHive/examples/invoker ~/vHive
~/vHive
INFO[2023-03-25T08:27:48.432587370Z] Reading the endpoints from the file: endpoints.json
INFO[2023-03-25T08:27:53.432888825Z] Issued / completed requests: 5, 0
INFO[2023-03-25T08:27:53.432922860Z] Real / target RPS: 0.00 / 1
INFO[2023-03-25T08:27:53.432936818Z] Experiment finished!
INFO[2023-03-25T08:27:53.432940883Z] The measured latencies are saved in rps0.00_lat.csv
```

No requests are completed...

**Comment by ustiugov** [comment]
@wangtianxia-sjtu thank you for the details. can I resolve the issue?

**Comment by wangtianxia-sjtu** [comment]
I am sorry but there is still something wrong...

In my previous [reply](https://github.com/vhive-serverless/vHive/issues/702#issuecomment-1483782949) above, I mentioned that when I met some error when I was running the command `./scripts/cluster/create_one_node_cluster.sh`. And the final test to invoke the function showed that none of the requests were completed. I attached all the logs in my reply above.

**Comment by ustiugov** [comment]
@wangtianxia-sjtu did I read it correctly that the problem is in image pulling? could you check if you can pull these images manually with docker?

**Comment by wangtianxia-sjtu** [comment]
I can pull the images with a proxy, via setting `http_proxy` and `https_proxy` with systems in docker.
By the way, the first error line is not about the image pulling, it is about the cri runtime interface.

```
[WARNING CRI]: container runtime is not running: output: time="2023-03-25T09:29:39Z" level=fatal msg="validate service connection: CRI v1 runtime API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService"
, error: exit status 1
```

Did you encounter similar errors before? The `vhive` program will listen on `/etc/vhive-cri/vhive-cri.sock`, but it seems it cannot respond correctly.😭

**Comment by ustiugov** [comment]
@CuriousGeorgiy  have you experienced this kind of error recently?

**Comment by Kingdo777** [comment]
> I can pull the images with a proxy, via setting `http_proxy` and `https_proxy` with systems in docker. By the way, the first error line is not about the image pulling, it is about the cri runtime interface.
> 
> ```
> [WARNING CRI]: container runtime is not running: output: time="2023-03-25T09:29:39Z" level=fatal msg="validate service connection: CRI v1 runtime API is not implemented for endpoint \"unix:///etc/vhive-cri/vhive-cri.sock\": rpc error: code = Unimplemented desc = unknown service runtime.v1.RuntimeService"
> , error: exit status 1
> ```
> 
> Did you encounter similar errors before? The `vhive` program will listen on `/etc/vhive-cri/vhive-cri.sock`, but it seems it cannot respond correctly.😭

我遇到了完全相同的问题，请问您现在已经解决了吗

**Comment by ustiugov** [comment]
@wangtianxia-sjtu I am very sorry for the late response. we found that the problem with pulling images is not a problem but a warning; the system should still work after that. 

Have you tried what I mentioned in my comment above?
>the fact that function invocations do not complete by the time invoker exits may be attributed to cold starts. To address the problem you can increase the experiment duration.

also, we have recently upgraded the main vHive dependencies (K8s, Knative, etc), so we would need you to reproduce it again. I apologize for the inconvenience.
